### ISSUE TITLE
Remove backwards in insert mode is broken

### ISSUE THREAD

**Issue by iamevie** [tauraamui/lilly#113]
When I try to remove the character on the right of my cursor in insert mode using the remove key, Lilly won't remove it but kind of bug out. Below I've attached a video of me using remove in Neovim and in Lilly to demonstrate what I mean.

https://github.com/tauraamui/lilly/assets/99840003/d6425fe5-fb57-4af5-ac42-e2088758e293

**Comment by tauraamui** [comment]
Do you mean the DELETE key? :)

**Comment by iamevie** [comment]
Yes that's the one. In German, it's called the remove key.
Nevertheless, is it suppose to work?

**Comment by tauraamui** [comment]
> Yes that's the one. In German, it's called the remove key. Nevertheless, is it suppose to work?

No, specifically because it doesn't work, it's supposed to be disabled. Well, I say disabled but really what I mean is we've disabled the key as found on ISO standard layout keyboards. What's actually happening when you enter keys which don't technically represent a normal visible character (either ANSI or UTF-8) that we're not handling the functionality of, it's really inserting a bunch of values as UTF-8 code points into the document which don't make sense, and causing whitespace to appear and the cursor to freak out.

**Comment by tauraamui** [comment]
Can you please try this example project in the V lang repo: https://github.com/vlang/v/blob/master/examples/term.ui/event_viewer.v and provide a list of the output if you press practically every key?

**Comment by iamevie** [comment]
> No, specifically because it doesn't work, it's supposed to be disabled. Well, I say disabled but really what I mean is we've disabled the key as found on ISO standard layout keyboards. What's actually happening when you enter keys which don't technically represent a normal visible character (either ANSI or UTF-8) that we're not handling the functionality of, it's really inserting a bunch of values as UTF-8 code points into the document which don't make sense, and causing whitespace to appear and the cursor to freak out.

I'm not using a standard ISO layout, which is probably why it still inserts the UTF-8 bytes. That's on me, but good to know!

**Comment by iamevie** [comment]
![image](https://github.com/tauraamui/lilly/assets/99840003/ab67178d-650b-42f3-a0d5-ad7fd77b5e5f)

Here you go :)

**Comment by valxntine** [comment]
> ![image](https://private-user-images.githubusercontent.com/99840003/288202547-ab67178d-650b-42f3-a0d5-ad7fd77b5e5f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE4MTAzMjYsIm5iZiI6MTcwMTgxMDAyNiwicGF0aCI6Ii85OTg0MDAwMy8yODgyMDI1NDctYWI2NzE3OGQtNjUwYi00MmYzLWEwZDUtYWQ3ZmQ3N2I1ZTVmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA1VDIxMDAyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk3YzBhODljM2Q5OWRlNDc3NTQ1ZGFkN2FjODk5NzYwM2VhODYzY2U4ZTcxYjdiZjQ2MDgyZDk1Y2ZlZGE5YTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.57dbYa4C02qqc3f31EHi-4PpGCI5zKlTG1F68Tfe6tQ)
> 
> Here you go :)

I believe it also captures mouse movements so if you press the key then move your mouse it'll change the event captured, if you're able to run it again, press your delete key then press escape and that should give us the proper event

**Comment by iamevie** [comment]
I've managed to capture the right key 
![image](https://github.com/tauraamui/lilly/assets/99840003/f3aad19c-55bd-49a9-a0d6-8c61d5bff120)
In general, it may be good to know that I don't use a normal keyboard but instead this [one](https://www.zsa.io/moonlander). In addition, I don't use qwertz or qwerty but colemak as a layout.

**Comment by tauraamui** [comment]
> I've managed to capture the right key ![image](https://private-user-images.githubusercontent.com/99840003/288206824-f3aad19c-55bd-49a9-a0d6-8c61d5bff120.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE4MTE3MDgsIm5iZiI6MTcwMTgxMTQwOCwicGF0aCI6Ii85OTg0MDAwMy8yODgyMDY4MjQtZjNhYWQxOWMtNTViZC00OWE5LWEwZDYtOGM2MWQ1YmZmMTIwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA1VDIxMjMyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJjMzFhZTYyYzhhMzc0YjJiNTZiZTI5ODBiMzEyMzAzNGNlODgwYmMwMjg3YjBjYTgyZWQ0NTEwMzM2N2ZhYjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.hgFpgPjl18WG-FvEoSMNfpzvVE0rQYpGTFfH_DOhDXQ) In general, it may be good to know that I don't use a normal keyboard but instead this [one](https://www.zsa.io/moonlander). In addition, I don't use qwertz or qwerty but colemak as a layout.

That's very helpful, thank you. That's interesting. In order to solve this issue, what I think we should really do is just have better support/understanding of different key layouts and board layouts and how to map between them. Let me go away and think of something to send you which we can use to send you to gather some data regarding your setup.

**Comment by iamevie** [comment]
> Basically, the functionality regarding key information we're receiving from term.ui currently is not good enough. It's identified your key's code as `null` which is not an identifier we can use. Therefore we will have to do some manual extra work to identify your keyboard's layout and do some internal mapping or similar.

Is that something I could do or that could be automated using a script? Something along the lines of 'oh you have a custom layout? Please press these keys in order to configure Lilly'

**Comment by tauraamui** [comment]
> > Basically, the functionality regarding key information we're receiving from term.ui currently is not good enough. It's identified your key's code as `null` which is not an identifier we can use. Therefore we will have to do some manual extra work to identify your keyboard's layout and do some internal mapping or similar.
> 
> Is that something I could do or that could be automated using a script? Something along the lines of 'oh you have a custom layout? Please press these keys in order to configure Lilly'

It's something which would ask X11 or Wayland, or (Windows UI?) whatever is being used as the window manager for the current layout and we would just need to have a set of mappings for each known one. For example, if I run `localectl list-x11-keymap-layouts | tr -s '\n' ',' | pbcopy` on my machine this tells me there's approx 99 different layouts known to X11.

```
af,al,am,ara,at,au,az,ba,bd,be,bg,br,brai,bt,bw,by,ca,cd,ch,
cm,cn,cz,de,dk,dz,ee,epo,es,et,fi,fo,fr,gb,ge,gh,gn,gr,hr,hu,
id,ie,il,in,iq,ir,is,it,jp,jv,ke,kg,kh,kr,kz,la,latam,lk,lt,lv,ma,mao,
md,me,mk,ml,mm,mn,mt,mv,my,nec_vndr/jp,ng,nl,no,np,ph,
pk,pl,pt,ro,rs,ru,se,si,sk,sn,sy,tg,th,tj,tm,tr,tw,tz,ua,us,uz,vn,za,
```
Having said that, I doubt Neovim is doing something similar to this, but I have yet to do enough research and reading to check. If Neovim is doing some magical easier technique realistically we should be doing the same. This is the first time I'm trying to solve a problem such as this.
